src/main/java/org/checkerframework/checker/crypto/CryptoAnnotatedTypeFactory.java

```diff
@@ -1,57 +1,123 @@
 
+    /** The @{@link AllowedAlgorithms} annotation. */
+    protected final AnnotationMirror ALLOWEDALGORITHMS =
+            AnnotationBuilder.fromClass(elements, AllowedAlgorithms.class);
+    /** The @{@link AllowedProviders} annotation. */
+    protected final AnnotationMirror ALLOWPROVIDERS =
+            AnnotationBuilder.fromClass(elements, AllowedProviders.class);
+    /** The @{@link Bottom} annotation. */
+    protected final AnnotationMirror BOTTOM = AnnotationBuilder.fromClass(elements, Bottom.class);
+    /** The @{@link UnknownAlgorithmOrProvider} annotation. */
+    protected final AnnotationMirror UNKNOWNALGORITHMORPROVIDER =
+            AnnotationBuilder.fromClass(elements, UnknownAlgorithmOrProvider.class);
+
     public CryptoAnnotatedTypeFactory(BaseTypeChecker checker) {
         super(checker);
         this.postInit();
     }
 
     @Override
-    public QualifierHierarchy createQualifierHierarchy(MultiGraphFactory factory) {
-        return new CryptoQualifierHierarchy(
-                factory, AnnotationBuilder.fromClass(elements, Bottom.class));
+    protected QualifierHierarchy createQualifierHierarchy() {
+        return new CryptoQualifierHierarchy(this.getSupportedTypeQualifiers(), elements);
     }
 
-    private static final class CryptoQualifierHierarchy extends GraphQualifierHierarchy {
+    private final class CryptoQualifierHierarchy extends MostlyNoElementQualifierHierarchy {
 
-        CryptoQualifierHierarchy(MultiGraphFactory factory, AnnotationMirror bottom) {
-            super(factory, bottom);
+        /** Qualifier kind for the @{@link AllowedAlgorithms} annotation. */
+        private final QualifierKind ALLOWEDALGORITHMS_KIND;
+        /** Qualifier kind for the @{@link AllowedProviders} annotation. */
+        private final QualifierKind ALLOWPROVIDERS_KIND;
+        /**
+         * Creates a CryptoQualifierHierarchy from the given classes.
+         *
+         * @param qualifierClasses classes of annotations that are the qualifiers for this hierarchy
+         * @param elements element utils
+         */
+        public CryptoQualifierHierarchy(
+                Collection<Class<? extends Annotation>> qualifierClasses, Elements elements) {
+            super(qualifierClasses, elements);
+            ALLOWEDALGORITHMS_KIND = getQualifierKind(ALLOWEDALGORITHMS);
+            ALLOWPROVIDERS_KIND = getQualifierKind(ALLOWPROVIDERS);
         }
 
         @Override
-        public boolean isSubtype(final AnnotationMirror subtype, final AnnotationMirror supertype) {
-            if (AnnotationUtils.areSameByClass(supertype, UnknownAlgorithmOrProvider.class)
-                    || AnnotationUtils.areSameByClass(subtype, Bottom.class)) {
+        public boolean isSubtypeWithElements(
+                AnnotationMirror subAnno,
+                QualifierKind subKind,
+                AnnotationMirror superAnno,
+                QualifierKind superKind) {
+            /* old */
```

Comment:
What does this mean?
Are some of these cases no longer needed? Is `isSubtypeWithElements` only called if the annotations actually have elements?
Should this use the qualifier kinds instead of checking the `.class`?

Reply:
I forgot to delete /*old*/.

Let me check the usage of qualifier kinds.